crates/isograph_schema/src/validate_argument_types.rs

```diff
@@ -266,30 +276,161 @@ pub fn value_satisfies_type<TOutputFormat: OutputFormat>(
+                        ValidateSchemaError::ExpectedTypeFoundObject {
+                            expected: id_annotation_to_typename_annotation(
+                                field_argument_definition_type,
+                                schema_data,
+                            ),
+                        },
+                        selection_supplied_argument_value.location,
+                    )),
+                    SelectionType::Object(object_id) => object_satisfies_type(
+                        selection_supplied_argument_value,
+                        variable_definitions,
+                        schema_data,
+                        server_fields,
+                        object_literal,
+                        object_id,
+                    ),
+                },
             }
         }
     }
 }
 
+fn object_satisfies_type<TOutputFormat: OutputFormat>(
+    selection_supplied_argument_value: &WithLocation<NonConstantValue>,
+    variable_definitions: &[WithSpan<
+        VariableDefinition<SelectionType<ServerObjectId, ServerScalarId>>,
+    >],
+    schema_data: &ServerFieldData<TOutputFormat>,
+    server_fields: &[ValidatedSchemaServerField<TOutputFormat>],
+    object_literal: &[NameValuePair<ValueKeyName, NonConstantValue>],
+    object_id: ServerObjectId,
+) -> Result<(), WithLocation<ValidateSchemaError>> {
+    let object = schema_data.object(object_id);
+    validate_no_extraneous_fields(
+        &object.encountered_fields,
+        object_literal,
+        selection_supplied_argument_value.location,
+    )?;
+
+    let missing_fields = get_missing_and_provided_fields(server_fields, object_literal, object)
+        .iter()
+        .filter_map(|field| match field {
+            ObjectLiteralFieldType::Provided(
+                field_type_annotation,
+                selection_supplied_argument_value,
+            ) => match value_satisfies_type(
+                &selection_supplied_argument_value.value,
+                field_type_annotation,
+                variable_definitions,
+                schema_data,
+                server_fields,
+            ) {
+                Ok(_) => None,
+                Err(e) => Some(Err(e)),
+            },
+            ObjectLiteralFieldType::Missing(field_name) => Some(Ok(*field_name)),
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    if missing_fields.is_empty() {
+        Ok(())
+    } else {
+        Err(WithLocation::new(
+            ValidateSchemaError::MissingFields {
+                missing_fields_names: missing_fields,
+            },
+            selection_supplied_argument_value.location,
+        ))
+    }
+}
+
+enum ObjectLiteralFieldType {
+    Provided(
+        GraphQLTypeAnnotation<SelectableServerFieldId>,
+        NameValuePair<ValueKeyName, NonConstantValue>,
+    ),
+    Missing(SelectableFieldName),
+}
+
+fn get_missing_and_provided_fields<TOutputFormat: OutputFormat>(
```

Comment:
nit: can we add a comment indicating that fields that are missing and not nullable are not returned from this? or we find a way to phrase it, e.g. (thanks DeepSeek!) `get_non_nullable_missing_and_provided_fields`